### DISCUSSION TITLE
Implementing new methods which treat incoming and rehearsal data separately

### DISCUSSION THREAD

**Discussion by pclucas14** [ContinualAI/avalanche#644] [Q&A] [answered]
Hi,

I'm trying to integrate [a new method ](https://arxiv.org/pdf/2104.05025.pdf) in your wonderful library. Our method treats incoming and rehearsal data differently. We apply logit masking on the incoming data and process the rehearsal data as is commonly done. Looking at the current code, my understanding is the following : 

1) Replay plugins overwrite the dataloader by outputting samples from the concatenation of buffer and incoming data
2) In this current setup, the strategy has no obvious way of telling the two data streams apart, (expect by looking directly at the label values)

In your opinion, what would be the easiest way to design a new strategy which could receive data from the incoming stream and the buffer _as two separate streams ?_ . My goal is to favor code reusability and hopefully piggyback on what is already implemented. 

Thank you!

**Comment by vlomonaco** [comment, marked as answer]
Hi @pclucas14! Thanks for taking avalanche into consideration for developing your strategy! I'm moving this to the Q&A discussion tab, as it seems more related to a usability question rather than a software issue. Let me know in case you want to move it back in the issues.

**Reply by pclucas14** [reply]
great, thanks Vincenzo

**Reply by AntonioCarta** [reply]
The easiest way is probably to treat as a regularization strategy instead of a replay strategy. This way, it's easier to separate the data and buffer streams.

you can use `before_backward` (you need a plugin for this, look at EWC for an example) to add your regularization terms:
```
def before_backward(...):
    # sample buffer mbatch
    # compute replay loss
    strategy.loss += self.lambda* buffer_loss
```

Also, replay strategies are made of DataLoaders and storage policies. You can use these components instead of reimplementing replay.

Hope this helps.

**Reply by vlomonaco** [reply]
@pclucas14 marking this as answered, let us know if you want to re-open it :)
